src/query/functions/aggregation/count_values.go

```diff
@@ -0,0 +1,268 @@
+	return countValuesOp{
+		params: params,
+		opType: opType,
+	}
+}
+
+type countValuesNode struct {
+	op         countValuesOp
+	controller *transform.Controller
+}
+
+// bucketColumn represents a column of times a particular value in a series has
+// been seen. This may expand as more unique values are seen
+type bucketColumn []int
+
+// bucketBlock is an abstraction for a set of series grouped by tags; count_values
+// works on these groupings rather than the entire set of series.
+type bucketBlock struct {
+	// columnLength can expand as further columns are processed; used to initialize
+	// the columns with empty values at each step
+	columnLength int
+	// columns indicates the number of times a value has been seen at a given step
+	columns []bucketColumn
+	// indexMapping maps any unique values seen to the appropriate column index
+	indexMapping map[float64]int
+}
+
+// Processes all series in this block bucket at the current column.
+func processBlockBucketAtColumn(
+	currentBucketBlock *bucketBlock,
+	values []float64,
+	bucket []int,
+	columnIndex int,
+) {
+	// Generate appropriate number of rows full of -1s that will later map to NaNs
+	// unless updated with valid values
+	currentColumnLength := currentBucketBlock.columnLength
+	currentBucketBlock.columns[columnIndex] = make(bucketColumn, currentColumnLength)
+	for i := 0; i < currentColumnLength; i++ {
+		ts.MemsetInt(currentBucketBlock.columns[columnIndex], -1)
+	}
+
+	countedValues := countValuesFn(values, bucket)
+	for distinctValue, count := range countedValues {
+		currentBucketColumn := currentBucketBlock.columns[columnIndex]
+		if rowIndex, seen := currentBucketBlock.indexMapping[distinctValue]; seen {
+			// This value has already been seen at rowIndex in a previous column
+			// so add the current value to the appropriate row index.
+			currentBucketColumn[rowIndex] = count
+		} else {
+			// The column index needs to be created here already
+			// Add the count to the end of the bucket column
+			currentBucketBlock.columns[columnIndex] = append(currentBucketColumn, count)
+
+			// Add the distinctValue to the indexMapping
+			currentBucketBlock.indexMapping[distinctValue] = len(currentBucketColumn)
+		}
+	}
+
+	currentBucketBlock.columnLength = len(currentBucketBlock.columns[columnIndex])
+}
+
+// Process the block
+func (n *countValuesNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	params := n.op.params
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	buckets, metas := utils.GroupSeries(
+		params.MatchingTags,
+		params.Without,
+		n.op.opType,
+		seriesMetas,
+	)
+
+	stepCount := stepIter.StepCount()
+	intermediateBlock := make([]bucketBlock, len(buckets))
+	for i := range intermediateBlock {
+		intermediateBlock[i].columns = make([]bucketColumn, stepCount)
+		intermediateBlock[i].indexMapping = make(map[float64]int, 10)
```

Comment:
did we decide to move all numbers to a top level constant ? initAlloc ?

Reply:
Changed this to be a more reasonable initialization size